From 2a57888d05d26398d7621d28841c4b21e72ca4fa Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Tue, 21 Jul 2026 19:20:54 +0200
Subject: [PATCH 45/90] arm64: dts: rockchip: rk3399-pinephone-pro: Add battery
 charger support

Describe Pinephone battery and battery charger. Battery has internal
NTC R25=3kOhm. Table was generated by aproximated beta model constant
as measured in my fridge. :)

The input current limit comes from the Type-C port controller, so the
charger takes fusb302 as its power supply.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 .../dts/rockchip/rk3399-pinephone-pro.dts     | 138 ++++++++++++++++++
 1 file changed, 138 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts b/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
index 3f78611bfd1a..e7577f92cb44 100644
--- a/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3399-pinephone-pro.dts
@@ -51,6 +51,44 @@ button-down {
 		};
 	};
 
+	bat: battery {
+                compatible = "simple-battery";
+                device-chemistry = "lithium-ion-polymer";
+                voltage-min-design-microvolt = <3400000>;
+                voltage-max-design-microvolt = <4350000>;
+                energy-full-design-microwatt-hours = <10290400>;
+                charge-full-design-microamp-hours = <2708000>;
+                precharge-current-microamp = <120000>;
+                charge-term-current-microamp = <150000>;
+                constant-charge-current-max-microamp = <1500000>;
+                constant-charge-voltage-max-microvolt = <4300000>;
+                factory-internal-resistance-micro-ohms = <150000>;
+                resistance-temp-table = <20 150>;
+                ocv-capacity-celsius = <20>;
+                ocv-capacity-table-0 =
+			<4308000 100>,
+			<4217000 95>,
+			<4161000 90>,
+			<4110000 85>,
+			<4066000 80>,
+			<4007000 75>,
+			<3974000 70>,
+			<3937000 65>,
+			<3896000 60>,
+			<3853000 55>,
+			<3827000 50>,
+			<3807000 45>,
+			<3793000 40>,
+			<3780000 35>,
+			<3768000 30>,
+			<3756000 25>,
+			<3740000 20>,
+			<3716000 15>,
+			<3689000 10>,
+			<3675000 5>,
+			<3400000 0>;
+        };
+
 	backlight: backlight {
 		compatible = "pwm-backlight";
 		pwms = <&pwm0 0 50000 0>;
@@ -482,6 +520,106 @@ otg_switch: OTG_SWITCH {
 				regulator-name = "otg_switch";
 			};
 		};
+
+		charger {
+			compatible = "rockchip,rk818-charger";
+			power-supplies = <&fusb0>;
+			monitored-battery = <&bat>;
+
+			/*
+			 * Vendor-specific tunables that do not map to the
+			 * standard simple-battery binding.
+			 */
+			rockchip,sense-resistor-micro-ohms = <10000>;
+			rockchip,relax-entry-current-microamp = <300000>;
+			rockchip,relax-exit-current-microamp = <300000>;
+			rockchip,max-soc-offset-percent = <60>;
+			rockchip,feedback-temp-celsius = <105>;
+			rockchip,monitor-sec = <5>;
+			rockchip,ntc-table = <
+				(-20) 26459
+				(-19) 25009
+				(-18) 23649
+				(-17) 22372
+				(-16) 21174
+				(-15) 20048
+				(-14) 18990
+				(-13) 17996
+				(-12) 17060
+				(-11) 16180
+				(-10) 15352
+				(-9) 14571
+				(-8) 13836
+				(-7) 13143
+				(-6) 12489
+				(-5) 11873
+				(-4) 11291
+				(-3) 10742
+				(-2) 10223
+				(-1) 9733
+				0 9269
+				1 8831
+				2 8416
+				3 8024
+				4 7653
+				5 7301
+				6 6968
+				7 6652
+				8 6353
+				9 6069
+				10 5800
+				11 5544
+				12 5301
+				13 5071
+				14 4852
+				15 4644
+				16 4446
+				17 4258
+				18 4079
+				19 3909
+				20 3747
+				21 3593
+				22 3446
+				23 3306
+				24 3172
+				25 3045
+				26 2924
+				27 2808
+				28 2697
+				29 2592
+				30 2491
+				31 2395
+				32 2304
+				33 2216
+				34 2132
+				35 2052
+				36 1975
+				37 1902
+				38 1832
+				39 1765
+				40 1700
+				41 1639
+				42 1580
+				43 1523
+				44 1469
+				45 1417
+				46 1368
+				47 1320
+				48 1274
+				49 1230
+				50 1188
+				51 1148
+				52 1109
+				53 1072
+				54 1036
+				55 1001
+				56 968
+				57 937
+				58 906
+				59 877
+				60 848
+			>;
+		};
 	};
 
 	vdd_cpu_b: regulator@40 {
-- 
2.55.0

